akka-actor/src/main/scala/akka/actor/CoordinatedShutdown.scala

```diff
@@ -0,0 +1,325 @@
+                // note that tasks within same phase are performed in parallel
+                val recoverEnabled = phases(phase).recover
+                val result = Future.sequence(tasks.map { task ⇒
+                  try {
+                    val r = task.apply()
+                    if (recoverEnabled) r.recover {
+                      case NonFatal(e) ⇒
+                        log.warning("Task failed in phase [{}]: {}", phase, e.getMessage)
+                        Done
+                    }
+                    else r
+                  } catch {
+                    case NonFatal(e) ⇒
+                      // in case task.apply throws
+                      if (recoverEnabled) {
+                        log.warning("Task failed in phase [{}]: {}", phase, e.getMessage)
+                        Future.successful(Done)
+                      } else
+                        Future.failed(e)
+                  }
+                }).map(_ ⇒ Done)(ExecutionContexts.sameThreadExecutionContext)
+                val timeout = phases(phase).timeout
+                val deadline = Deadline.now + timeout
+                val timeoutFut = after(timeout, system.scheduler) {
+                  if (phase == CoordinatedShutdown.PhaseActorSystemTerminate && deadline.hasTimeLeft) {
+                    // too early, i.e. triggered by system termination
+                    result
+                  } else if (result.isCompleted)
+                    Future.successful(Done)
+                  else if (recoverEnabled) {
+                    log.warning("Coordinated shutdown phase [{}] timed out after {}", phase, timeout)
+                    Future.successful(Done)
+                  } else
+                    Future.failed(
+                      new TimeoutException(s"Coordinated shutdown phase [$phase] timed out after $timeout"))
+                }
+                Future.firstCompletedOf(List(result, timeoutFut))
+            })
+            if (remaining.isEmpty)
+              phaseResult // avoid flatMap when system terminated in last phase
+            else
+              phaseResult.flatMap(_ ⇒ loop(remaining))
+        }
+      }
+      val done = loop(orderedPhases)
+      runPromise.completeWith(done)
+    }
+    runPromise.future
+  }
+
+  /**
+   * The configured timeout for a given `phase`.
+   * For example useful as timeout when actor `ask` requests
+   * is used as a task.
+   */
+  def timeout(phase: String): FiniteDuration =
+    phases.get(phase) match {
+      case Some(Phase(_, timeout, _)) ⇒ timeout
+      case None ⇒
+        throw new IllegalArgumentException(s"Unknown phase [$phase]. All phases must be defined in configuration")
+    }
+
+  /**
+   * Sum of timeouts of all phases that have some task.
+   */
+  def totalTimeout(): FiniteDuration = {
+    import scala.collection.JavaConverters._
+    tasks.keySet.asScala.foldLeft(Duration.Zero) {
+      case (acc, phase) ⇒ acc + timeout(phase)
+    }
```

Comment:
This is not any exact measure, just a convenience to find out a decent timeout. The extra microseconds between each phase is neglectable since all phases will not use their full timeout.